File: C#/problem9.cs
using System;
using System.Diagnostics;

class Problem9
{
  public static void Main(string[] args)
  {
    Stopwatch stopwatch = new Stopwatch();
    stopwatch.Start();
    Console.WriteLine(SpecialPythagoreanTriplet());
    stopwatch.Stop();
    Console.WriteLine(stopwatch.ElapsedMilliseconds / 1000.0 + " seconds");
  }

  public static int SpecialPythagoreanTriplet()
  {
    int specialSum = 1000;
    int specialProduct = 0;
    for (int i = 1; i < specialSum - 1 && specialProduct == 0; i++)
    {
        for (int j = 1; j < specialSum - 1; j++)
        {
          int[] triplet = new int[] {i, j, specialSum - i - j};
          Array.Sort(triplet);
          if ((triplet[2] + triplet[1] + triplet[0] == specialSum) &&
              (Math.Pow(triplet[0], 2) + Math.Pow(triplet[1], 2) ==
              Math.Pow(triplet[2], 2)))
          {
            specialProduct = triplet[0] * triplet[1] * triplet[2];
          }
        }
    }
    return specialProduct;
  }
}

File: Ruby/problem9.rb
def special_pythagorean_triplet
  special_sum = 1000
  # need to leave room for at least 1 + 1 + (special_sum - 2)
  combinations = []
  for i in 1...special_sum -1 do
    for j in 1...special_sum -1 do
      combinations << [i, j, special_sum - i - j].sort if special_sum - i - j > 0
    end
  end
  unique_combinations = combinations.uniq
  triplet_index = -1
  unique_combinations.each.with_index do |combo, index|
    if combo[0] ** 2 + combo[1] ** 2 === combo[2] ** 2
      triplet_index = index
    end
  end
  unique_combinations[triplet_index]
end

# Product of triplet
start_time = Time.now
puts special_pythagorean_triplet.reduce(:*)
puts "#{Time.now - start_time} seconds."
